lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb

```diff
@@ -0,0 +1,202 @@
+require "topological_inventory/ansible_tower/collector"
+require "topological_inventory/providers/common/operations/sources_api_client"
+require "topological_inventory/ansible_tower/receptor/async_receiver"
+
+module TopologicalInventory
+  module AnsibleTower
+    module TargetedRefresh
+      class ServiceInstance < TopologicalInventory::AnsibleTower::Collector
+        REFS_PER_REQUEST_LIMIT = 20
+
+        def initialize(payload = {})
+          self.params    = payload['params']
+          self.source_id = payload['source_id']
+
+          # TODO: add metrics exporter
+          super(payload['source_uid'], nil)
+        end
+
+        # Entrypoint for 'ServiceInstance.refresh' operation
+        def refresh
+          return if params_missing?
+
+          # Connection settings
+          set_connection_data!
+          logger.info("ServiceInstance#refresh - Connection set successfully")
+
+          # Get all tasks
+          tasks, refresh_state_uuid = {}, SecureRandom.uuid
+          params.to_a.each do |task|
+            if task['task_id'].blank? || task['source_ref'].blank?
+              logger.warn("ServiceInstance#refresh - missing data for task: #{task}")
+              next
+            end
+            tasks[task['task_id']] = task['source_ref']
+
+            if tasks.length == REFS_PER_REQUEST_LIMIT
+              refresh_part(tasks, refresh_state_uuid, SecureRandom.uuid)
+              tasks = {}
+            end
+          end
+
+          refresh_part(tasks, refresh_state_uuid, SecureRandom.uuid) unless tasks.empty?
+        rescue => err
+          logger.error("ServiceInstance#refresh - Error: #{err.message}\n#{err.backtrace.join("\n")}")
+        end
+
+        private
+
+        attr_accessor :params, :source_id
+
+        def required_params
+          %w[source_id source params]
+        end
+
+        def params_missing?
+          is_missing = false
+          required_params.each do |attr|
+            if (is_missing = send(attr).blank?)
+              logger.error("ServiceInstance#refresh - Missing #{attr} for the availability_check request [Source ID: #{source_id}]")
+              break
+            end
+          end
+
+          is_missing
+        end
+
+        def refresh_part(tasks, refresh_state_uuid, refresh_state_part_uuid)
+          tasks_id = tasks.keys.join(' | id: ')
+
+          # API request, nodes and jobs under workflow job not needed
+          query_params = {
+            :id__in    => tasks.values.join(','),
+            :page_size => limits['service_instances']
+          }
+
+          if on_premise?
+            refresh_part_on_premise(tasks_id, refresh_state_uuid, query_params)
+          else
+            refresh_part_cloud(tasks_id, refresh_state_uuid, refresh_state_part_uuid, query_params)
+          end
+        end
+
+        def refresh_part_on_premise(_tasks_id, refresh_state_uuid, query_params)
```

Comment:
is there a reason `refresh_part_on_premise` is not public vs `refresh_part_cloud` is?

Reply:
both of them are private. Hm, obviously it's not a good idea to include public block inside private methods, I'll move it